apps/settings/js/modules/base/module.js

```diff
@@ -0,0 +1,150 @@
+ *
+ *   var instance = ExtendingModule();
+ *   instance.inc();
+ *   instance.print(); // 101
+ *
+ * @module modules/base/module
+ */
+define(function() {
+  'use strict';
+
+  var LOG_LEVEL = {
+    NONE: 0,
+    TRACE: 1,
+    DEBUG: 2,
+    INFO: 3,
+    WARN: 4,
+    ERROR: 5,
+    ALL: 6
+  };
+
+  var _constructorMap = (function() {
+    var _ctorMap = new Map();
+    var _getConstructor = function(Module) {
+      return _ctorMap.get(Module) || function() {};
+    };
+    var _registerConstructor = function(Module, constructor) {
+      _ctorMap.set(Module, constructor || function() {});
+    };
+    return {
+      getConstructor: _getConstructor,
+      registerConstructor: _registerConstructor
+    };
+  })();
+
+  var _emptyFunction = function() {};
+  var _createLogger = function(name) {
+    switch (name) {
+      case 'TRACE':
+      case 'DEBUG':
+      case 'INFO':
+        return function(msg) {
+          console.log(name + ' ' + this._msgPrefix + msg);
+        };
+      case 'WARN':
+        return function(msg) {
+          console.warn(name + ' ' + this._msgPrefix + msg);
+        };
+      case 'ERROR':
+        return function(msg) {
+          console.error(name + ' ' + this._msgPrefix + msg);
+        };
+    }
+  };
+
+  var ModulePrototype = {
+    get _msgPrefix() {
+      return '[' + this.$name + ']: ';
+    },
+    trace: _emptyFunction,
+    debug: _emptyFunction,
+    info: _emptyFunction,
+    warn: _emptyFunction,
+    error: _emptyFunction,
+    throw: function(msg) {
+      throw new Error(this._msgPrefix + msg);
+    },
+    set _logLevel(value) {
+      Object.keys(LOG_LEVEL).forEach((name) => {
+        var level = LOG_LEVEL[name];
+        if (value >= level && value > 0) {
+          this[name.toLowerCase()] = _createLogger(name);
+        } else {
+          this[name.toLowerCase()] = _emptyFunction;
+        }
+      });
+    },
+    super: _constructorMap.getConstructor
+  };
+
+  var _extend = function() {
+    switch (arguments.length) {
+      case 0:
+        break;
+      case 1:
+        var Module = arguments[0];
+        for (var i in Module.prototype) {
+          if (i === '$name') {
+            continue;
+          }
+          var pd = Object.getOwnPropertyDescriptor(Module.prototype, i);
+          if (pd) {
+            Object.defineProperty(this.prototype, i, pd);
+          }
+        }
+        break;
+      default:
+        Array.prototype.forEach.call(arguments, (Module) => {
+          _extend.call(this, Module);
+        });
+        break;
+    }
+    return this;
+  };
+
+  // Returns a module function. The function returns an instance of the
+  // module when it is called. The passed constructor is called using the
+  // instance as "this".
+  function _create(constructor) {
+    var ModuleFunc = function() {};
+    ModuleFunc.prototype = Object.create(ModulePrototype);
+    ModuleFunc.prototype.$name = constructor && constructor.name || '';
```

Comment:
This means that developers have to explicitly define the constructor function with a name, otherwise its name would always be `""`. They can override `xxx.$name` later of course when designing module, but I am not sure whether the `$` would be easy to understand or not.